### PR TITLE
Hw 8.descriptor

### PR DESCRIPTION
**Pull request by ninaprus**
-EmailDescriptor
-Singleton(type)
-IngegerField

class Product, Basket, Order

### CODE REVIEW

File: hw_7.context_manager.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        print(self.time)
+
+
+def my_func3():
+    sum([i for i in range(10000)])
+    return
+
+with timer():
+    my_func3()
+    print('Runtime: ', end='')
```
Comment (katevoskresenska) [inline hw_7.context_manager.py:76]: Pull request should contain files for the specific home work only

File: hw_9.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+# Реализовать синглтон метакласс(класс для создания классов синглтонов).
+
+class Singleton(type):
+    cache = {}
+    def __call__(instance, *args, **kwargs):
+        if instance not in Singleton.cache:
+            Singleton.cache[instance] = super().__call__(*args, **kwargs)
+            return Singleton.cache[instance]
+        else:
+            return Singleton.cache[instance]
```
Comment (katevoskresenska) [inline hw_9.py:40]: else can be omitted

    def __call__(instance, *args, **kwargs):
        if instance not in Singleton.cache:
            Singleton.cache[instance] = super().__call__(*args, **kwargs)
        return Singleton.cache[instance]

File: hw_9.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+
+    def __get__(self, instance, owner):
+        return instance._price * 1.2
+
+    def __set__(self, instance, value):
+        instance._price = value
+
+
+class Product:
+    _warehouse = {}
```
Comment (katevoskresenska) [inline hw_9.py:98]: Product and Warehouse should be different classes

File: hw_9.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+        return self._warehouse[self.name]
+
+    def get_availb(self):
+        if self.quantity > 0:
+            return True
+        else:
+            return False
+
+# 2) Добавить товар на склад
+    def add_to_warehouse(self, quantity):
```
Comment (mmykhailenko) [inline hw_9.py:124]: this method should be in Warehouse class

File: hw_9.py [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+
+    def _update_quantity(self, quantity):
+        self._quantity = quantity
+        return self._quantity
+
+    def get_total_price(self, quantity):
+        return quantity * self.price
+
+# 3) Удалить товар со склада
+    def remove_from_warehouse(self, quantity=None):
```
Comment (mmykhailenko) [inline hw_9.py:142]: same..in warehouse

File: hw_9.py [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+        print(cls._warehouse)
+
+# 7) Распечатать список товаров с заданной категорией
+    @classmethod
+    def get_all_product(cls):
+        print(cls.__name__, ':', [key for key in cls._warehouse])
+
+
+# 6) Товар может принадлежать к категории
+class Clothes(Product):
```
Comment (mmykhailenko) [inline hw_9.py:172]: not a good idea create a class per each category

File: hw_9.py [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+        self._basket[product.name] = {'quantity': self.quantity, 'price': self.total_price}
+
+    def check_quantity(self, product, quantity):
+        if product.get_availb:
+            if quantity <= product.quantity:
+                return quantity
+        return None
+
+    def get_total_price(self, product, quantity):
+        if self.quantity == 0:
```
Comment (mmykhailenko) [inline hw_9.py:193]: `if not self.quantity`

**Review by mmykhailenko — APPROVED** [review]

**Pull request merged**